README.md
## Gauged

[![tests][travis]][travis-builds]

An append-only storage layer for gauges, counters, timers and other numeric data types that change over time.

Features:

- Comfortably handle billions of data points on a single node.
- Support for sparse data, unlike the fixed-size RRDtool.
- Cache-oblivious data structures and algorithms for speed and memory-efficiency.
- Range queries and roll-ups of any size down to the configurable resolution of 1 second.
- Use either **MySQL**, **PostgreSQL** or **SQLite** for persistence.
- Runs on Mac, Linux & Windows.

## Installation

The library can be installed with **easy_install** or **pip**

```bash
$ pip install gauged
```

Python 2.7.x (e.g. CPython or PyPy) is required.

## Example

Writing data:

```python
from gauged import Gauged

gauged = Gauged('mysql://root@localhost/gauged')

with gauged.writer as writer:
    writer.add({ 'requests': 1, 'response_time': 0.45, 'memory_usage': 145.6 })
    writer.add({ 'requests': 1, 'response_time': 0.25, 'cpu_usage': 148.3, 'api_requests': 3 })
```

Reading data:

```python
# Count the total number of requests
requests = gauged.aggregate('requests', Gauged.SUM)

# Count the number of requests between 2014/01/01 and 2014/01/08
requests = gauged.aggregate('requests', Gauged.SUM, start=datetime(2014, 1, 1),
    end=datetime(2014, 1, 8))

# Get the 95th percentile response time from the past week
response_time = gauged.aggregate('response_time', Gauged.PERCENTILE,
    percentile=95, start=-Gauged.WEEK)

# Get latest memory usage
memory_usage = gauged.value('memory_usage')
```

Plotting data (using [matplotlib][matplotlib]):

```python
import pylab
series = gauged.aggregate_series('requests', gauged.SUM, interval=gauged.DAY,
    start=-gauged.WEEK)
pylab.plot(series.dates, series.values, label='Requests per day for the past week')
pylab.show()
```

## Documentation

See the [documentation][documentation] or [technical overview][technical-overview].

## Tests

Setup the test environment:

```bash
$ virtualenv env
$ source ./env/bin/activate
$ pip install -r test-requirements.txt
```

You can then run the test suite using an in-memory driver:

```bash
$ make check-quick
```

To run the full suite, first edit the configuration in `test_drivers.cfg` so that PostgreSQL and MySQL both point to existing (and empty) databases, then run:

```bash
$ make check
```

You can run coverage analysis with `make coverage` and run a lint tool `make lint`.

## Benchmarks

Use `make build` followed by `python benchmark [OPTIONS]` to run benchmarks using a SQLite-based in-memory database. YMMV.

**python benchmark.py --number 1000000 --days 365**

```
Writing to sqlite:// (block_size=86400000, resolution=1000)
Spreading 1M measurements to key "foobar" over 365 days
Wrote 1M measurements in 5.388 seconds (185.6K/s) (rss: 12.7MB)
Gauge data uses 7.6MB (8B per measurement)
min() in 0.024s (read 41.9M measurements/s) (rss: 12.8MB)
max() in 0.023s (read 43.4M measurements/s) (rss: 12.8MB)
sum() in 0.023s (read 42.7M measurements/s) (rss: 12.8MB)
count() in 0.024s (read 42.1M measurements/s) (rss: 12.8MB)
mean() in 0.028s (read 36.1M measurements/s) (rss: 12.8MB)
stddev() in 0.05s (read 20.1M measurements/s) (rss: 12.8MB)
median() in 0.06s (read 16.8M measurements/s) (rss: 27.7MB)
```

**python benchmark.py --number 100000000 --days 365**

```
Writing to sqlite:// (block_size=86400000, resolution=1000)
Spreading 100M measurements to key "foobar" over 365 days
Wrote 100M measurements in 405.925 seconds (246.4K/s) (rss: 21.7MB)
Gauge data uses 502.2MB (5.26601144B per measurement)
min() in 0.818s (read 122.3M measurements/s) (rss: 21.7MB)
max() in 0.79s (read 126.6M measurements/s) (rss: 21.7MB)
sum() in 0.785s (read 127.4M measurements/s) (rss: 21.7MB)
count() in 0.766s (read 130.5M measurements/s) (rss: 21.7MB)
mean() in 0.891s (read 112.3M measurements/s) (rss: 21.7MB)
stddev() in 1.697s (read 58.9M measurements/s) (rss: 21.7MB)
median() in 3.547s (read 28.2M measurements/s) (rss: 1007.9MB)
```

## License (MIT)

```
Copyright (c) 2014 Chris O'Hara <cohara87@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

[travis]: https://api.travis-ci.org/chriso/gauged.png?branch=master
[travis-builds]: https://travis-ci.org/chriso/gauged
[technical-overview]: https://github.com/chriso/gauged/blob/master/docs/technical-overview.md
[documentation]: https://github.com/chriso/gauged/blob/master/docs/documentation.md
[matplotlib]: http://matplotlib.org/


HISTORY
---[ 1.0.1 ]

* Fixed a bug when sorting large arrays
* PEP8 style fixes

---[ 1.0.0 ]

* Started following semver
* Added a script to migrate gauged schemas
* Delete all data for a key before or after a timestamp

---[ 0.4.1 ]

* Fix linux builds when installed via pip

---[ 0.4.0 ]

* Windows support

---[ 0.3.0 ]

* Fixed a bug with the mean aggregate
* Improved test coverage
* Reduced pedantic compiler warnings

---[ 0.2.1 ]

* Fixed a minor memory leak

---[ 0.2.0 ]

* Reduce the memory requirements of stddev(), percentile() and median()
* More aggressive caching for associative aggregates
* Remove setuptools dependency
* Fixed the driver querystring parser

---[ 0.1.0 ]

* Initial release


benchmark.py
"""
Gauged - https://github.com/chriso/gauged
Copyright 2014 (c) Chris O'Hara <cohara87@gmail.com>
"""

from random import random
from math import floor, ceil
from time import time
from calendar import timegm
from datetime import datetime, timedelta
from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter
from gauged import Gauged


def abbreviate(suffixes, cutoff):
    def abbreviate_(number, decimals=1):
        position = 0
        while position < len(suffixes) and abs(number) >= cutoff:
            number = round(number / cutoff, decimals)
            position += 1
        if floor(number) == ceil(number):
            number = int(number)
        return str(number) + suffixes[position]
    return abbreviate_

abbreviate_number = abbreviate(['', 'K', 'M', 'B'], 1000)
abbreviate_bytes = abbreviate(['B', 'KB', 'MB', 'GB', 'TB'], 1024)

# Parse CLI options
benchmark = ArgumentParser(usage='%(prog)s [OPTIONS]',
                           formatter_class=ArgumentDefaultsHelpFormatter)
benchmark.add_argument(
    '-n', '--number', type=int, default=1000000,
    help='How many measurements to store')
benchmark.add_argument(
    '-t', '--days', type=int, default=365,
    help='How many days to spread the measurements over')
benchmark.add_argument(
    '-d', '--driver', default='sqlite://',
    help='Where to store the data (defaults to SQLite in-memory)')
benchmark.add_argument(
    '-b', '--block-size', type=int, default=Gauged.DAY,
    help='The block size to use')
benchmark.add_argument(
    '-r', '--resolution', type=int, default=Gauged.SECOND,
    help='The resolution to use')
options = vars(benchmark.parse_args())

# Setup the Gauged instance
gauged = Gauged(options['driver'], block_size=options['block_size'],
                resolution=options['resolution'], key_overflow=Gauged.IGNORE,
                gauge_nan=Gauged.IGNORE)
gauged.sync()

print 'Writing to %s (block_size=%s, resolution=%s)' % \
    (options['driver'], options['block_size'], options['resolution'])

# Get the start and end timestamp
end = datetime.now()
start = end - timedelta(days=options['days'])
start_timestamp = timegm(start.timetuple())
end_timestamp = timegm(end.timetuple())

number = abbreviate_number(options['number'])

print 'Spreading %s measurements to key "foobar" over %s days' % \
    (number, options['days'])

# Benchmark writes
measurements = options['number']
span = end_timestamp - start_timestamp
start = time()
with gauged.writer as writer:
    data = ['foobar', 0]
    gauges = [data]
    add = writer.add
    for timestamp in xrange(start_timestamp, end_timestamp,
                            span // measurements):
        data[1] = random()
        add(gauges, timestamp=timestamp*1000)
elapsed = time() - start

print 'Wrote %s measurements in %s seconds (%s/s)' % \
    (number, round(elapsed, 3), abbreviate_number(measurements / elapsed))

statistics = gauged.statistics()
byte_count = statistics.byte_count
print 'Gauge data uses %s (%s per measurement)' % \
    (abbreviate_bytes(byte_count),
     abbreviate_bytes(byte_count / float(measurements)))

# Read benchmarks
for aggregate in ('min', 'max', 'sum', 'count', 'mean', 'stddev', 'median'):
    start = time()
    gauged.aggregate('foobar', aggregate)
    elapsed = time() - start
    print '%s() in %ss (read %s measurements/s)' % \
        (aggregate, round(elapsed, 3),
         abbreviate_number(measurements / elapsed))
